browser/misc_metrics/vertical_tab_metrics.cc

```diff
@@ -0,0 +1,189 @@
+#include "chrome/browser/ui/tabs/tab_group_model.h"
+#include "chrome/browser/ui/tabs/tab_strip_model.h"
+#include "components/prefs/pref_registry_simple.h"
+#include "components/prefs/pref_service.h"
+
+namespace misc_metrics {
+
+namespace {
+
+const int kOpenTabsBuckets[] = {1, 5, 10, 50};
+const int kGroupAndPinnedTabsBuckets[] = {2, 5};
+
+const char* GetHistogramNameForCountType(TabCountType count_type) {
+  switch (count_type) {
+    case TabCountType::kOpen:
+      return kVerticalOpenTabsHistogramName;
+    case TabCountType::kGroup:
+      return kVerticalGroupTabsHistogramName;
+    case TabCountType::kPinned:
+      return kVerticalPinnedTabsHistogramName;
+  }
+}
+
+const char* GetStoragePrefNameForCountType(TabCountType count_type) {
+  switch (count_type) {
+    case TabCountType::kOpen:
+      return kMiscMetricsOpenTabsStorage;
+    case TabCountType::kGroup:
+      return kMiscMetricsGroupTabsStorage;
+    case TabCountType::kPinned:
+      return kMiscMetricsPinnedTabsStorage;
+  }
+}
+
+void RecordMaxToHistogramBucket(TabCountType count_type, uint64_t max_value) {
+  const char* histogram_name = GetHistogramNameForCountType(count_type);
+  switch (count_type) {
+    case TabCountType::kOpen:
+      p3a_utils::RecordToHistogramBucket(histogram_name, kOpenTabsBuckets,
+                                         max_value);
+      break;
+    case TabCountType::kGroup:
+    case TabCountType::kPinned:
+      p3a_utils::RecordToHistogramBucket(histogram_name,
+                                         kGroupAndPinnedTabsBuckets, max_value);
+      break;
+  }
+}
+
+}  // namespace
+
+const char kVerticalOpenTabsHistogramName[] = "Brave.VerticalTabs.OpenTabs";
+const char kVerticalGroupTabsHistogramName[] = "Brave.VerticalTabs.GroupTabs";
+const char kVerticalPinnedTabsHistogramName[] = "Brave.VerticalTabs.PinnedTabs";
+
+VerticalTabBrowserMetrics::VerticalTabBrowserMetrics(
+    PrefService* profile_prefs,
+    base::RepeatingClosure change_callback)
+    : profile_prefs_(profile_prefs), change_callback_(change_callback) {
+  for (TabCountType count_type : kAllTabCountTypes) {
+    counts_[count_type] = 0;
+  }
+  pref_change_registrar_.Init(profile_prefs);
+  pref_change_registrar_.Add(
+      brave_tabs::kVerticalTabsEnabled,
+      base::BindRepeating(&VerticalTabBrowserMetrics::UpdateEnabledStatus,
+                          base::Unretained(this)));
+  UpdateEnabledStatus();
+}
+
+VerticalTabBrowserMetrics::~VerticalTabBrowserMetrics() = default;
+
+void VerticalTabBrowserMetrics::OnTabStripModelChanged(
+    TabStripModel* tab_strip_model,
+    const TabStripModelChange& change,
+    const TabStripSelectionChange& selection) {
+  if (!vertical_tabs_enabled_) {
+    return;
+  }
+  counts_[TabCountType::kOpen] = tab_strip_model->count();
+  TabGroupModel* group_model = tab_strip_model->group_model();
+  if (group_model != nullptr) {
+    counts_[TabCountType::kGroup] = group_model->ListTabGroups().size();
+  }
```

Comment:
I'm curious when this can happen as well. Like other place, I'd prefer CHECK() if it's unlike to happen but we want to make it sure.

Reply:
If the `TabGroupModelFactory` is not provided to the TabStripModel upon construction, the group_model will be nullptr: https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ui/tabs/tab_strip_model.cc;drc=e5cf5b446e32349e986ceca1c4f9c374db878a14;l=291

Reply:
Thanks for the reference. I didn't know that tab groups can be disabled 😄